File: Algorithm/LintCode/BredthFirstSearch/CourseSchedule2.cpp
#include <vector>
#include <queue>
#include <map>

using namespace std;
class Solution {
public:
    /*
     * @param numCourses: a total of n courses
     * @param prerequisites: a list of prerequisite pairs
     * @return: the course order
     */
    vector<int> findOrder(int numCourses, vector<pair<int, int>> &prerequisites) {
        // 1. build a graph
        vector<int> degrees(numCourses, 0);
        map<int, vector<int>> graph;
        
        // build graph
        for(const pair<int, int>& p : prerequisites){
            int courseNum = p.first;
            int preReq = p.second;
            
            // treat as directed graph [1,0], direction is 0 -> 1, so 1's degree += 1
            graph[preReq].push_back(courseNum);
            ++degrees[courseNum];
        }
        
        
        queue<int> myQueue;
        
        // push all 0 degree course Number to a std::queue
        for(int i = 0; i < numCourses; ++i){
            if(degrees[i] == 0){
                myQueue.push(i);
            }
        }
        
        
        // need a count to see if there is a topological sortt
        int count = 0;
        vector<int> out;
        
        // do a topological sort
        while(!myQueue.empty()){
            
            int courseId = myQueue.front();
            myQueue.pop();
            out.push_back(courseId);
            ++count;
            
            for(int i : graph[courseId]){
                degrees[i] -= 1;
                
                if(degrees[i] == 0){
                    myQueue.push(i);
                }
            }
        }
        
        if(count == numCourses){
            //there is a possible way to finish all courses
            return out;
        }
        
        
        return vector<int>();
    }
};

File: Algorithm/LintCode/BredthFirstSearch/CourseSchedule2.py

"""
616. Course Schedule II
https://www.lintcode.com/problem/course-schedule-ii
There are a total of n courses you have to take, labeled from 0 to n - 1.
Some courses may have prerequisites, for example to take course 0 you have to first take course 1, which is expressed as a pair: [0,1]
Given the total number of courses and a list of prerequisite pairs, return the ordering of courses you should take to finish all courses.
There may be multiple correct orders, you just need to return one of them. If it is impossible to finish all courses, return an empty array.

Example 1:
Input: n = 2, prerequisites = [[1,0]] 
Output: [0,1]

Example 2:
Input: n = 4, prerequisites = [[1,0],[2,0],[3,1],[3,2]] 
Output: [0,1,2,3] or [0,2,1,3]

弃用计数，改用一个list来存储纪录
"""

from collections import deque

class Solution:
    """
    @param: numCourses: a total of n courses
    @param: prerequisites: a list of prerequisite pairs
    @return: the course order
    """
    def findOrder(self, numCourses, prerequisites):
        #init graph and degree
        graph = {i: [] for i in range(numCourses)}
        degree = [0 for i in range(numCourses)]
        
        # build graph
        for course, pre in prerequisites:
            graph[pre].append(course)
            degree[course] += 1
            
        my_queue = deque()
        course_path = []
        
        #push all 0 degree preRequests
        for i in range(numCourses):
            if degree[i] == 0:
                my_queue.append(i)
                           
        while my_queue:
            course = my_queue.popleft()
            course_path.append(course)
            
            for c in graph[course]:
                degree[c] -= 1
                if degree[c] == 0:
                    my_queue.append(c)
                     
        if len(course_path) == numCourses:
            return course_path
        
        
        return []